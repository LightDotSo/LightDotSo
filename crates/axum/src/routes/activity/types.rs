// Copyright (C) 2023 Light, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU Affero General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Affero General Public License for more details.
//
// You should have received a copy of the GNU Affero General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.

use lightdotso_prisma::activity;
use serde::{Deserialize, Serialize};
use utoipa::ToSchema;

// -----------------------------------------------------------------------------
// Types
// -----------------------------------------------------------------------------

/// Activity root type.
#[derive(Serialize, Deserialize, ToSchema, Clone)]
#[serde(rename_all = "snake_case")]
pub(crate) struct Activity {
    /// The id of the activity.
    id: String,
    /// The entity id of the activity.
    entity: String,
    /// The operation type of the activity.
    operation: String,
    /// The timestamp of the activity.
    timestamp: String,
    /// The actor wallet address of the activity.
    address: Option<String>,
}

/// Implement From<activity::Data> for Activity.
impl From<activity::Data> for Activity {
    fn from(activity: activity::Data) -> Self {
        Self {
            id: activity.id,
            entity: activity.entity.to_string(),
            operation: activity.operation.to_string(),
            timestamp: activity.timestamp.to_rfc3339(),
            address: activity.wallet_address.map(|addr| addr.to_string()),
        }
    }
}
