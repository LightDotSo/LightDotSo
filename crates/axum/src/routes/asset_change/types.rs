// Copyright 2023-2024 LightDotSo.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

use crate::routes::{interpretation_action::types::InterpretationAction, token::types::Token};
use lightdotso_prisma::asset_change;
use serde::{Deserialize, Serialize};
use utoipa::ToSchema;

// -----------------------------------------------------------------------------
// Types
// -----------------------------------------------------------------------------

/// AssetChange root type.
#[derive(Serialize, Deserialize, ToSchema, Clone)]
#[serde(rename_all = "snake_case")]
pub(crate) struct AssetChange {
    /// The id of the asset change.
    id: String,
    /// The address of the asset change.
    address: String,
    /// The amount of the asset change.
    amount: i64,
    /// The before amount of the asset change.
    before_amount: Option<i64>,
    /// The after amount of the asset change.
    after_amount: Option<i64>,
    /// The action id of the asset change.
    action: Option<InterpretationAction>,
    /// The token id of the asset change.
    token: Option<Token>,
}

// -----------------------------------------------------------------------------
// From
// -----------------------------------------------------------------------------

/// Implement From<asset_change::Data> for AssetChange.
impl From<asset_change::Data> for AssetChange {
    fn from(asset_change: asset_change::Data) -> Self {
        Self {
            id: asset_change.id,
            address: asset_change.address,
            amount: asset_change.amount.parse().unwrap_or_default(),
            before_amount: asset_change.before_amount.and_then(|s| s.parse::<i64>().ok()),
            after_amount: asset_change.after_amount.and_then(|s| s.parse::<i64>().ok()),
            action: asset_change.interpretation_action.and_then(|maybe_action| {
                maybe_action.map(|action| InterpretationAction::from(*action))
            }),
            token: asset_change
                .token
                .and_then(|maybe_token| maybe_token.map(|token| Token::from(*token))),
        }
    }
}
