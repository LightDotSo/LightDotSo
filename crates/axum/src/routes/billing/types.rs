// Copyright (C) 2023 Light, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU Affero General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Affero General Public License for more details.
//
// You should have received a copy of the GNU Affero General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.

use lightdotso_prisma::billing;
use serde::{Deserialize, Serialize};
use utoipa::ToSchema;

// -----------------------------------------------------------------------------
// Types
// -----------------------------------------------------------------------------

/// Billing root type.
#[derive(Serialize, Deserialize, ToSchema, Clone)]
#[serde(rename_all = "snake_case")]
pub(crate) struct Billing {
    /// The id of the billing.
    id: String,
}

// -----------------------------------------------------------------------------
// From
// -----------------------------------------------------------------------------

/// Implement From<billing::Data> for Billing.
impl From<billing::Data> for Billing {
    fn from(billing: billing::Data) -> Self {
        Self { id: billing.id }
    }
}
