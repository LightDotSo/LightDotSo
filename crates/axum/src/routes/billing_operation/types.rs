// Copyright 2023-2024 LightDotSo.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

use lightdotso_prisma::billing_operation;
use serde::{Deserialize, Serialize};
use utoipa::ToSchema;

// -----------------------------------------------------------------------------
// Types
// -----------------------------------------------------------------------------

/// BillingOperation root type.
#[derive(Serialize, Deserialize, ToSchema, Clone)]
#[serde(rename_all = "snake_case")]
pub(crate) struct BillingOperation {
    /// The id of the billing operation.
    id: String,
    /// The status of the billing operation.
    status: String,
    /// The balance USD of the billing operation.
    balance_usd: f64,
}

// -----------------------------------------------------------------------------
// From
// -----------------------------------------------------------------------------

/// Implement From<billing_operation::Data> for BillingOperation.
impl From<billing_operation::Data> for BillingOperation {
    fn from(billing_operation: billing_operation::Data) -> Self {
        Self {
            id: billing_operation.id,
            status: billing_operation.status.to_string(),
            balance_usd: billing_operation.balance_usd,
        }
    }
}
