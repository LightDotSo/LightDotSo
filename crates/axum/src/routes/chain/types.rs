// Copyright (C) 2023 Light, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU Affero General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Affero General Public License for more details.
//
// You should have received a copy of the GNU Affero General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.

use lightdotso_prisma::chain;
use serde::{Deserialize, Serialize};
use utoipa::ToSchema;

// -----------------------------------------------------------------------------
// Types
// -----------------------------------------------------------------------------

/// Chain root type.
#[derive(Serialize, Deserialize, ToSchema, Clone)]
#[serde(rename_all = "snake_case")]
pub(crate) struct Chain {
    /// The id of the protocol group.
    id: i64,
}

// -----------------------------------------------------------------------------
// From
// -----------------------------------------------------------------------------

/// Implement From<chain::Data> for Chain.
impl From<chain::Data> for Chain {
    fn from(chain: chain::Data) -> Self {
        Self { id: chain.id }
    }
}
