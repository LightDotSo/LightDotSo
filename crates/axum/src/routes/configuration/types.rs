// Copyright (C) 2023 Light, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU Affero General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Affero General Public License for more details.
//
// You should have received a copy of the GNU Affero General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.

use lightdotso_prisma::{configuration, owner};
use serde::{Deserialize, Serialize};
use utoipa::ToSchema;

// -----------------------------------------------------------------------------
// Types
// -----------------------------------------------------------------------------

/// Configuration root type.
#[derive(Serialize, Deserialize, ToSchema, Clone)]
pub(crate) struct Configuration {
    /// The id of the configuration.
    id: String,
    /// The address of the configuration.
    address: String,
    /// The image hash of the configuration.
    image_hash: String,
    /// The checkpoint of the configuration.
    checkpoint: i64,
    /// The threshold of the configuration.
    threshold: i64,
    /// The owners of the configuration.
    owners: Vec<ConfigurationOwner>,
}

/// Configuration Owner.
#[derive(Serialize, Deserialize, ToSchema, Clone)]
pub(crate) struct ConfigurationOwner {
    /// The id of the owner.
    id: String,
    /// The index of the owner.
    index: i32,
    /// The address of the owner.
    address: String,
    /// The weight of the owner.
    weight: i64,
}

/// Implement From<configuration::Data> for Configuration.
impl From<configuration::Data> for Configuration {
    fn from(configuration: configuration::Data) -> Self {
        Self {
            id: configuration.id.to_string(),
            address: configuration.address.to_string(),
            image_hash: configuration.image_hash.to_string(),
            checkpoint: configuration.checkpoint,
            threshold: configuration.threshold,
            owners: configuration.owners.map_or(Vec::new(), |owners| {
                owners.into_iter().map(ConfigurationOwner::from).collect()
            }),
        }
    }
}

/// Implement From<owner::Data> for Owner.
impl From<owner::Data> for ConfigurationOwner {
    fn from(owner: owner::Data) -> Self {
        Self {
            id: owner.id.to_string(),
            address: owner.address.to_string(),
            index: owner.index,
            weight: owner.weight,
        }
    }
}
