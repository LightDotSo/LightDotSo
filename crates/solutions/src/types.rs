// Copyright (C) 2023 Light, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU Affero General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Affero General Public License for more details.
//
// You should have received a copy of the GNU Affero General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.

use ethers::{
    types::{Address, H256},
    utils::hex,
};
use serde::{
    de::{self, Visitor},
    Deserialize, Deserializer, Serialize, Serializer,
};
use serde_with::serde_as;
use std::convert::TryFrom;

#[derive(Clone, Debug, PartialEq)]
pub struct Signature(pub Vec<u8>);

/// The struct representation of a wallet signer
/// Derived from: https://github.com/0xsequence/go-sequence/blob/eabca0c348b5d87dd943a551908c80f61c347899/config.go#L17
/// License: Apache-2.0
#[derive(Clone, Debug, Serialize, Deserialize, PartialEq)]
pub struct Signer {
    pub weight: u8,
    pub leaf: SignatureLeaf,
}

#[derive(Clone, Debug, Serialize, Deserialize, PartialEq)]
pub struct SignerNode {
    #[serde(skip_serializing_if = "Option::is_none")]
    pub signer: Option<Signer>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub left: Option<Box<SignerNode>>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub right: Option<Box<SignerNode>>,
}

/// The struct representation of a wallet config
/// Derived from: https://github.com/0xsequence/go-sequence/blob/eabca0c348b5d87dd943a551908c80f61c347899/config.go#L12
/// License: Apache-2.0
#[derive(Clone, Debug, Serialize, Deserialize, PartialEq)]
pub struct WalletConfig {
    // Bytes32 hash of the checkpoint
    pub checkpoint: u32,
    // Uint16 threshold
    pub threshold: u16,
    // Uint256 weight of the retured signature
    pub weight: usize,
    // Image hash of the wallet config that is used to verify the wallet
    pub image_hash: H256,
    // Signers of the wallet
    pub tree: SignerNode,
    // Internal field used to store the image hash of the wallet config
    #[serde(skip_serializing_if = "Option::is_none")]
    pub internal_root: Option<H256>,
}

#[derive(Clone, Debug, Serialize, Deserialize, PartialEq)]
pub enum SignatureLeaf {
    ECDSASignature(ECDSASignatureLeaf),
    AddressSignature(AddressSignatureLeaf),
    DynamicSignature(DynamicSignatureLeaf),
    NodeSignature(NodeLeaf),
    BranchSignature(BranchLeaf),
    SubdigestSignature(SubdigestLeaf),
    NestedSignature(NestedLeaf),
}

/// The enum representation of a signature leaf type
/// Derived from: https://github.com/0xsequence/wallet-contracts/blob/e0c5382636a88b4db4bcf0a70623355d7cd30fb4/contracts/modules/commons/submodules/auth/SequenceBaseSig.sol#L102
/// License: Apache-2.0
#[derive(Clone, Debug, Serialize, Deserialize, PartialEq)]
#[repr(u8)]
pub enum SignatureLeafType {
    ECDSASignature = 0,
    Address = 1,
    DynamicSignature = 2,
    Node = 3,
    Branch = 4,
    Subdigest = 5,
    Nested = 6,
}

/// The struct representation of an ECDSA signature leaf type
/// Derived from: https://github.com/0xsequence/wallet-contracts/blob/e0c5382636a88b4db4bcf0a70623355d7cd30fb4/contracts/utils/SignatureValidator.sol#L83
/// License: Apache-2.0
#[derive(Clone, Debug, Serialize, Deserialize, PartialEq)]
#[repr(u8)]
pub enum ECDSASignatureType {
    ECDSASignatureTypeEIP712 = 1,
    ECDSASignatureTypeEthSign = 2,
}

/// The constant length of an ECDSA signature
/// The actual length of the signature is 65 bytes
/// Internally, the original length is + 1 byte for the signature type
pub const ECDSA_SIGNATURE_LENGTH: usize = 65;

pub const ERC1271_MAGICVALUE_BYTES32: [u8; 4] = [22, 38, 186, 126];

#[derive(Clone, Debug, PartialEq)]
pub struct ECDSASignature(pub [u8; ECDSA_SIGNATURE_LENGTH]);

#[serde_as]
#[derive(Clone, Debug, Serialize, Deserialize, PartialEq)]
pub struct ECDSASignatureLeaf {
    pub address: Address,
    pub signature_type: ECDSASignatureType,
    pub signature: ECDSASignature,
}

#[serde_as]
#[derive(Clone, Debug, Serialize, Deserialize, PartialEq)]
pub struct AddressSignatureLeaf {
    pub address: Address,
}

/// The struct representation of a Dynamic signature leaf type
#[derive(Clone, Debug, Serialize, Deserialize, PartialEq)]
#[repr(u8)]
pub enum DynamicSignatureType {
    DynamicSignatureTypeEIP712 = 1,
    DynamicSignatureTypeEthSign = 2,
    DynamicSignatureTypeEIP1271 = 3,
}

#[derive(Clone, Debug, Serialize, Deserialize, PartialEq)]
pub struct DynamicSignatureLeaf {
    pub address: Address,
    pub signature_type: DynamicSignatureType,
    pub signature: Signature,
}

#[derive(Clone, Debug, Serialize, Deserialize, PartialEq)]
pub struct NodeLeaf {}

#[derive(Clone, Debug, Serialize, Deserialize, PartialEq)]
pub struct BranchLeaf {}

#[derive(Clone, Debug, Serialize, Deserialize, PartialEq)]
pub struct SubdigestLeaf {}

#[derive(Clone, Debug, Serialize, Deserialize, PartialEq)]
pub struct NestedLeaf {
    pub internal_threshold: u16,
    pub external_weight: u8,
    pub address: Address,
    pub internal_root: H256,
}

impl Signature {
    pub fn len(&self) -> usize {
        let Signature(inner) = self;
        inner.len()
    }

    pub fn is_empty(&self) -> bool {
        self.len() == 0
    }

    pub fn as_slice(&self) -> &[u8] {
        let Signature(inner) = self;
        inner.as_slice()
    }
}

impl From<Vec<u8>> for Signature {
    fn from(bytes: Vec<u8>) -> Self {
        Signature(bytes)
    }
}

impl From<[u8; ECDSA_SIGNATURE_LENGTH]> for ECDSASignature {
    fn from(item: [u8; ECDSA_SIGNATURE_LENGTH]) -> Self {
        ECDSASignature(item)
    }
}

impl TryFrom<Vec<u8>> for ECDSASignature {
    type Error = &'static str; // You can use a more specific error type here

    fn try_from(bytes: Vec<u8>) -> Result<Self, Self::Error> {
        if bytes.len() != ECDSA_SIGNATURE_LENGTH {
            return Err("Invalid length for ECDSASignature");
        }

        let mut array = [0u8; ECDSA_SIGNATURE_LENGTH];
        for (place, element) in array.iter_mut().zip(bytes.iter()) {
            *place = *element;
        }
        Ok(ECDSASignature(array))
    }
}

impl Serialize for ECDSASignature {
    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: Serializer,
    {
        let hex_string = format!("0x{}", hex::encode(self.0));
        serializer.serialize_str(&hex_string)
    }
}

struct ECDSASignatureVisitor;

impl<'de> Visitor<'de> for ECDSASignatureVisitor {
    type Value = ECDSASignature;

    fn expecting(&self, formatter: &mut std::fmt::Formatter) -> std::fmt::Result {
        formatter.write_str("a string starts with 0x followed by 260 hexadecimal characters")
    }

    fn visit_str<E>(self, value: &str) -> Result<Self::Value, E>
    where
        E: de::Error,
    {
        if !value.starts_with("0x") {
            return Err(E::custom("Expected string to start with '0x'"));
        }
        let bytes = hex::decode(&value[2..]).map_err(de::Error::custom)?;
        if bytes.len() != ECDSA_SIGNATURE_LENGTH {
            return Err(E::custom("Expected 65 bytes"));
        }
        let mut array = [0u8; ECDSA_SIGNATURE_LENGTH];
        array.copy_from_slice(&bytes);
        Ok(ECDSASignature(array))
    }
}

impl<'de> Deserialize<'de> for ECDSASignature {
    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
    where
        D: Deserializer<'de>,
    {
        deserializer.deserialize_str(ECDSASignatureVisitor)
    }
}

struct SignatureVisitor;

impl<'de> Visitor<'de> for SignatureVisitor {
    type Value = Signature;

    fn expecting(&self, formatter: &mut std::fmt::Formatter) -> std::fmt::Result {
        formatter.write_str("a string starts with 0x followed by hexadecimal characters")
    }

    fn visit_str<E>(self, value: &str) -> Result<Self::Value, E>
    where
        E: de::Error,
    {
        if !value.starts_with("0x") {
            return Err(E::custom("Expected string to start with '0x'"));
        }
        let bytes = hex::decode(&value[2..]).map_err(de::Error::custom)?;
        Ok(Signature(bytes))
    }
}

impl<'de> Deserialize<'de> for Signature {
    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
    where
        D: Deserializer<'de>,
    {
        deserializer.deserialize_str(SignatureVisitor)
    }
}

impl Serialize for Signature {
    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: Serializer,
    {
        let hex_string = format!("0x{}", hex::encode(self.0.as_slice()));
        serializer.serialize_str(&hex_string)
    }
}
