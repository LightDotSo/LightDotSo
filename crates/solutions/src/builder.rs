// Copyright (C) 2023 Light, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU Affero General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Affero General Public License for more details.
//
// You should have received a copy of the GNU Affero General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.

// Entire file's examples and configurations from: https://github.com/0xsequence/sequence.js/blob/e5659ab1a304ae48b28c843b0d99fb3b3f6bc0b1/packages/core/tests/v2/config.spec.ts#L4
// License: Apache-2.0
// Truly thank you for providing this out there as free and open source work.
// We are forever grateful

use crate::types::SignerNode;
use eyre::{eyre, Result};

// From: https://github.com/0xsequence/sequence.js/blob/e5659ab1a304ae48b28c843b0d99fb3b3f6bc0b1/packages/core/src/v2/config.ts#L317
// License: Apache-2.0
// Roots the nodes seuqentially in order.
pub fn rooted_node_builder(members: Vec<SignerNode>) -> Result<SignerNode> {
    if members.is_empty() {
        return Err(eyre!("Empty members vector"));
    }

    let mut acc = members[0].clone();
    for member in members.iter().skip(1) {
        acc = SignerNode {
            signer: None,
            left: Some(Box::new(acc)),
            right: Some(Box::new(member.clone())),
        }
    }
    Ok(acc)
}

#[cfg(test)]
mod tests {
    use crate::{
        types::{AddressSignatureLeaf, NestedLeaf, SignatureLeaf, Signer, SubdigestLeaf},
        utils::parse_hex_to_bytes32,
    };

    use super::*;

    #[test]
    fn test_config_1() -> Result<()> {
        let members = vec![
            SignerNode {
                signer: Some(Signer {
                    weight: Some(2),
                    leaf: SignatureLeaf::AddressSignature(AddressSignatureLeaf {
                        address: "0x07ab71Fe97F9122a2dBE3797aa441623f5a59DB1".parse()?,
                    }),
                }),
                left: None,
                right: None,
            },
            SignerNode {
                signer: Some(Signer {
                    weight: None,
                    leaf: SignatureLeaf::SubdigestSignature(SubdigestLeaf {
                        hash: parse_hex_to_bytes32(
                            "0xb374baf809e388014912ca7020c8ef51ad68591db3f010f9e35a77c15d4d6bed",
                        )?
                        .into(),
                    }),
                }),
                left: None,
                right: None,
            },
            SignerNode {
                signer: Some(Signer {
                    weight: None,
                    leaf: SignatureLeaf::SubdigestSignature(SubdigestLeaf {
                        hash: parse_hex_to_bytes32(
                            "0x787c83a19321fc70f8653f8faa39cce60bf26cac51c25df1b0634eb7ddbe0c60",
                        )?
                        .into(),
                    }),
                }),
                left: None,
                right: None,
            },
            SignerNode {
                signer: Some(Signer {
                    weight: Some(1),
                    leaf: SignatureLeaf::AddressSignature(AddressSignatureLeaf {
                        address: "0xdafea492d9c6733ae3d56b7ed1adb60692c98bc5".parse()?,
                    }),
                }),
                left: None,
                right: None,
            },
        ];

        let node = rooted_node_builder(members)?;
        insta::assert_debug_snapshot!(node);

        Ok(())
    }

    #[test]
    fn test_config_2() -> Result<()> {
        let members = vec![
            SignerNode {
                signer: Some(Signer {
                    weight: None,
                    leaf: SignatureLeaf::SubdigestSignature(SubdigestLeaf {
                        hash: parse_hex_to_bytes32(
                            "0x0000000000000000000000000000000000000000000000000000000000000000",
                        )?
                        .into(),
                    }),
                }),
                left: None,
                right: None,
            },
            SignerNode {
                signer: Some(Signer {
                    weight: None,
                    leaf: SignatureLeaf::SubdigestSignature(SubdigestLeaf {
                        hash: parse_hex_to_bytes32(
                            "0x0000000000000000000000000000000000000000000000000000000000000001",
                        )?
                        .into(),
                    }),
                }),
                left: None,
                right: None,
            },
            SignerNode {
                signer: Some(Signer {
                    weight: None,
                    leaf: SignatureLeaf::SubdigestSignature(SubdigestLeaf {
                        hash: parse_hex_to_bytes32(
                            "0x0000000000000000000000000000000000000000000000000000000000000002",
                        )?
                        .into(),
                    }),
                }),
                left: None,
                right: None,
            },
            SignerNode {
                signer: Some(Signer {
                    weight: None,
                    leaf: SignatureLeaf::SubdigestSignature(SubdigestLeaf {
                        hash: parse_hex_to_bytes32(
                            "0x0000000000000000000000000000000000000000000000000000000000000003",
                        )?
                        .into(),
                    }),
                }),
                left: None,
                right: None,
            },
            SignerNode {
                signer: Some(Signer {
                    weight: None,
                    leaf: SignatureLeaf::SubdigestSignature(SubdigestLeaf {
                        hash: parse_hex_to_bytes32(
                            "0x0000000000000000000000000000000000000000000000000000000000000004",
                        )?
                        .into(),
                    }),
                }),
                left: None,
                right: None,
            },
            SignerNode {
                signer: Some(Signer {
                    weight: None,
                    leaf: SignatureLeaf::SubdigestSignature(SubdigestLeaf {
                        hash: parse_hex_to_bytes32(
                            "0x0000000000000000000000000000000000000000000000000000000000000005",
                        )?
                        .into(),
                    }),
                }),
                left: None,
                right: None,
            },
            SignerNode {
                signer: Some(Signer {
                    weight: None,
                    leaf: SignatureLeaf::SubdigestSignature(SubdigestLeaf {
                        hash: parse_hex_to_bytes32(
                            "0x0000000000000000000000000000000000000000000000000000000000000006",
                        )?
                        .into(),
                    }),
                }),
                left: None,
                right: None,
            },
            SignerNode {
                signer: Some(Signer {
                    weight: None,
                    leaf: SignatureLeaf::SubdigestSignature(SubdigestLeaf {
                        hash: parse_hex_to_bytes32(
                            "0x0000000000000000000000000000000000000000000000000000000000000007",
                        )?
                        .into(),
                    }),
                }),
                left: None,
                right: None,
            },
        ];

        let node = rooted_node_builder(members)?;
        insta::assert_debug_snapshot!(node);

        Ok(())
    }

    #[test]
    fn test_config_3() -> Result<()> {
        let members = vec![
            SignerNode {
                signer: Some(Signer {
                    weight: None,
                    leaf: SignatureLeaf::NestedSignature(NestedLeaf {
                        internal_root: [0; 32].into(),
                        internal_threshold: 0,
                        external_weight: 3,
                        size: 3,
                    }),
                }),
                left: Some(Box::new(
                    SignerNode {
                        signer: None,
                        left: Some(Box::new(
                            SignerNode {
                                signer: Some(Signer {
                                    weight: Some(2),
                                    leaf: SignatureLeaf::AddressSignature(AddressSignatureLeaf {
                                        address: "0x07ab71Fe97F9122a2dBE3797aa441623f5a59DB1".parse()?,
                                    }),
                                }),
                                left: None,
                                right: None,
                            },
                        )),
                        right: Some(Box::new(
                            SignerNode {
                                signer: Some(Signer {
                                    weight: None,
                                    leaf: SignatureLeaf::SubdigestSignature(SubdigestLeaf {
                                        hash: parse_hex_to_bytes32(
                                            "0x0000000000000000000000000000000000000000000000000000000000000006",
                                        )
                                        ?
                                        .into(),
                                    }),
                                }),
                                left: None,
                                right: None,
                            },
                        )),
                    },
                )),
                right: Some(Box::new(
                    SignerNode {
                        signer: Some(Signer {
                            weight: None,
                            leaf: SignatureLeaf::SubdigestSignature(SubdigestLeaf {
                                hash: parse_hex_to_bytes32(
                                    "0x787c83a19321fc70f8653f8faa39cce60bf26cac51c25df10000000000000000",
                                )
                                ?
                                .into(),
                            }),
                        }),
                        left: None,
                        right: None,
                    },
                )),
            },
            SignerNode {
                signer: Some(Signer {
                    weight: None,
                    leaf: SignatureLeaf::SubdigestSignature(SubdigestLeaf {
                        hash: parse_hex_to_bytes32(
                            "0xb374baf809e388014912ca7020c8ef51ad68591db3f010f9e35a77c15d4d6bed",
                        )
                        ?
                        .into(),
                    }),
                }),
                left: None,
                right: None,
            },
            SignerNode {
                signer: Some(Signer {
                    weight: None,
                    leaf: SignatureLeaf::SubdigestSignature(SubdigestLeaf {
                        hash: parse_hex_to_bytes32(
                            "0x787c83a19321fc70f8653f8faa39cce60bf26cac51c25df1b0634eb7ddbe0c60",
                        )
                        ?
                        .into(),
                    }),
                }),
                left: None,
                right: None,
            },
            SignerNode {
                signer: Some(Signer {
                    weight: Some(1),
                    leaf: SignatureLeaf::AddressSignature(AddressSignatureLeaf {
                        address: "0xdafea492d9c6733ae3d56b7ed1adb60692c98bc5".parse()?,
                    }),
                }),
                left: None,
                right: None,
            },
        ];

        let node = rooted_node_builder(members)?;
        insta::assert_debug_snapshot!(node);

        Ok(())
    }
}
