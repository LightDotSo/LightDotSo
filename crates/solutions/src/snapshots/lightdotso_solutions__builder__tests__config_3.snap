---
source: crates/solutions/src/builder.rs
expression: node
---
SignerNode {
    signer: None,
    left: Some(
        SignerNode {
            signer: None,
            left: Some(
                SignerNode {
                    signer: None,
                    left: Some(
                        SignerNode {
                            signer: Some(
                                Signer {
                                    weight: None,
                                    leaf: NestedSignature(
                                        NestedLeaf {
                                            internal_threshold: 0,
                                            external_weight: 3,
                                            internal_root: 0x0000000000000000000000000000000000000000000000000000000000000000,
                                            size: 3,
                                        },
                                    ),
                                },
                            ),
                            left: Some(
                                SignerNode {
                                    signer: None,
                                    left: Some(
                                        SignerNode {
                                            signer: Some(
                                                Signer {
                                                    weight: Some(
                                                        2,
                                                    ),
                                                    leaf: AddressSignature(
                                                        AddressSignatureLeaf {
                                                            address: 0x07ab71fe97f9122a2dbe3797aa441623f5a59db1,
                                                        },
                                                    ),
                                                },
                                            ),
                                            left: None,
                                            right: None,
                                        },
                                    ),
                                    right: Some(
                                        SignerNode {
                                            signer: Some(
                                                Signer {
                                                    weight: None,
                                                    leaf: SubdigestSignature(
                                                        SubdigestLeaf {
                                                            hash: 0x0000000000000000000000000000000000000000000000000000000000000006,
                                                        },
                                                    ),
                                                },
                                            ),
                                            left: None,
                                            right: None,
                                        },
                                    ),
                                },
                            ),
                            right: Some(
                                SignerNode {
                                    signer: Some(
                                        Signer {
                                            weight: None,
                                            leaf: SubdigestSignature(
                                                SubdigestLeaf {
                                                    hash: 0x787c83a19321fc70f8653f8faa39cce60bf26cac51c25df10000000000000000,
                                                },
                                            ),
                                        },
                                    ),
                                    left: None,
                                    right: None,
                                },
                            ),
                        },
                    ),
                    right: Some(
                        SignerNode {
                            signer: Some(
                                Signer {
                                    weight: None,
                                    leaf: SubdigestSignature(
                                        SubdigestLeaf {
                                            hash: 0xb374baf809e388014912ca7020c8ef51ad68591db3f010f9e35a77c15d4d6bed,
                                        },
                                    ),
                                },
                            ),
                            left: None,
                            right: None,
                        },
                    ),
                },
            ),
            right: Some(
                SignerNode {
                    signer: Some(
                        Signer {
                            weight: None,
                            leaf: SubdigestSignature(
                                SubdigestLeaf {
                                    hash: 0x787c83a19321fc70f8653f8faa39cce60bf26cac51c25df1b0634eb7ddbe0c60,
                                },
                            ),
                        },
                    ),
                    left: None,
                    right: None,
                },
            ),
        },
    ),
    right: Some(
        SignerNode {
            signer: Some(
                Signer {
                    weight: Some(
                        1,
                    ),
                    leaf: AddressSignature(
                        AddressSignatureLeaf {
                            address: 0xdafea492d9c6733ae3d56b7ed1adb60692c98bc5,
                        },
                    ),
                },
            ),
            left: None,
            right: None,
        },
    ),
}
